File: src/server/backup_engine.py
import asyncio
import os
import tarfile
import tempfile
from datetime import datetime

import boto3
import structlog

from config.provider import GlobalConfigProvider

logger = structlog.get_logger()


class BackupEngine:
    """Automated Disaster Recovery Engine for Point-in-Time Recovery (PITR)"""

    def __init__(self):
        self.running = False
        self._task = None
        self.data_dir = "./data"

    def start(self):
        """Starts the background backup daemon."""
        config = GlobalConfigProvider().get_config()
        if not hasattr(config, "backups") or not config.backups.enabled:
            logger.info("Automated backups disabled via config.toml")
            return

        self.running = True
        self._task = asyncio.create_task(self._backup_loop())
        logger.info(
            "Backup Engine started", interval_minutes=config.backups.interval_minutes
        )

    async def stop(self):
        """Gracefully stops the backup daemon."""
        self.running = False
        if self._task:
            self._task.cancel()
            try:
                await self._task
            except asyncio.CancelledError:
                pass
        logger.info("Backup Engine stopped")

    async def _backup_loop(self):
        while self.running:
            config = GlobalConfigProvider().get_config()
            interval = config.backups.interval_minutes * 60

            try:
                # Wait for the interval in small increments to allow responsive shutdown
                slept = 0
                while slept < interval and self.running:
                    await asyncio.sleep(1)
                    slept += 1

                if not self.running:
                    break

                await self._execute_backup(config.backups)
            except asyncio.CancelledError:
                break
            except Exception as e:
                logger.error("Backup Engine encountered fatal error", error=str(e))
                await asyncio.sleep(60)  # Wait before retry

    async def _execute_backup(self, backup_config):
        """Compresses the data directory and uploads to S3."""
        if not os.path.exists(self.data_dir):
            logger.warning("Backup skipped: Data directory does not exist", path=self.data_dir)
            return

        timestamp = datetime.utcnow().strftime("%Y%m%d_%H%M%S")
        archive_name = f"axiom_backup_{timestamp}.tar.gz"
        tmp_dir = tempfile.gettempdir()
        archive_path = os.path.join(tmp_dir, archive_name)

        try:
            logger.info("Starting automated PITR backup...", dest=archive_path)

            # 1. Compress data directory
            def make_tarfile():
                with tarfile.open(archive_path, "w:gz") as tar:
                    tar.add(self.data_dir, arcname=os.path.basename(self.data_dir))

            await asyncio.to_thread(make_tarfile)

            # 2. Upload to S3
            if not backup_config.s3_access_key or not backup_config.s3_secret_key:
                logger.warning("Backup S3 credentials missing. Skipping upload.")
                return

            client_args = {
                "service_name": "s3",
                "region_name": backup_config.s3_region,
                "aws_access_key_id": backup_config.s3_access_key,
                "aws_secret_access_key": backup_config.s3_secret_key,
            }
            if backup_config.s3_endpoint_url:
                client_args["endpoint_url"] = backup_config.s3_endpoint_url

            def upload():
                s3_client = boto3.client(**client_args)
                s3_client.upload_file(
                    archive_path, backup_config.s3_bucket, archive_name
                )

            await asyncio.to_thread(upload)
            logger.info(
                "Backup uploaded to S3 successfully",
                archive=archive_name,
                bucket=backup_config.s3_bucket,
            )

        except Exception as e:
            logger.error("Failed to execute S3 backup", error=str(e))
        finally:
            if os.path.exists(archive_path):
                try:
                    os.remove(archive_path)
                except Exception as ex:
                    logger.warning("Failed to clean up temp backup archive", error=str(ex))


backup_engine = BackupEngine()

File: src/server/backup_engine.rs
use crate::config::loader::ConfigManager;
use aws_sdk_s3::config::Region;
use aws_sdk_s3::Client;
use flate2::write::GzEncoder;
use flate2::Compression;
use std::fs::File;
use std::path::Path;
use std::time::Duration;
use tar::Builder;
use tokio::time::sleep;

pub struct BackupEngine;

impl BackupEngine {
    pub fn start() -> tokio::task::JoinHandle<()> {
        tokio::spawn(async move {
            loop {
                let config = ConfigManager::get();
                if !config.backups.enabled {
                    break;
                }

                let interval_secs = (config.backups.interval_minutes * 60) as u64;
                sleep(Duration::from_secs(interval_secs)).await;

                if let Err(e) = Self::execute_backup().await {
                    eprintln!("Backup failed: {}", e);
                }
            }
        })
    }

    async fn execute_backup() -> Result<(), Box<dyn std::error::Error>> {
        let config = ConfigManager::get();
        let data_dir = Path::new("./data");

        if !data_dir.exists() {
            println!("Backup skipped: data directory does not exist.");
            return Ok(());
        }

        let timestamp = chrono::Utc::now().format("%Y%m%d_%H%M%S").to_string();
        let archive_name = format!("axiom_backup_{}.tar.gz", timestamp);
        let tmp_dir = Path::new("./.tmp_backup");
        std::fs::create_dir_all(tmp_dir)?;

        let archive_path = tmp_dir.join(&archive_name);

        // 1. Compress
        let tar_gz = File::create(&archive_path)?;
        let enc = GzEncoder::new(tar_gz, Compression::default());
        let mut tar = Builder::new(enc);
        tar.append_dir_all("data", data_dir)?;
        tar.finish()?;

        // 2. Upload
        if config.backups.s3_access_key.is_empty() || config.backups.s3_secret_key.is_empty() {
            println!("S3 credentials missing. Skipping upload.");
            return Ok(());
        }

        std::env::set_var("AWS_ACCESS_KEY_ID", &config.backups.s3_access_key);
        std::env::set_var("AWS_SECRET_ACCESS_KEY", &config.backups.s3_secret_key);

        let region = Region::new(config.backups.s3_region.clone());
        let mut aws_config_builder = aws_config::defaults(aws_config::BehaviorVersion::latest()).region(region);

        if let Some(endpoint) = &config.backups.s3_endpoint_url {
            aws_config_builder = aws_config_builder.endpoint_url(endpoint);
        }

        let sdk_config = aws_config_builder.load().await;
        let client = Client::new(&sdk_config);

        let body = aws_sdk_s3::primitives::ByteStream::from_path(&archive_path).await?;

        client
            .put_object()
            .bucket(&config.backups.s3_bucket)
            .key(&archive_name)
            .body(body)
            .send()
            .await?;

        println!("Backup uploaded to S3 successfully: {}", archive_name);

        // Cleanup
        let _ = std::fs::remove_file(archive_path);

        Ok(())
    }
}
